### DISCUSSION TITLE
After v1.15.6, how to get source line / line number associated to instruction pointer ?

### DISCUSSION THREAD

**Discussion by buixor** [antonmedv/expr#544] [Q&A] [answered]
Hello, 

With (I believe) [v1.15.6](https://github.com/expr-lang/expr/releases/tag/v1.15.6), some fields of `vm.Program` aren't accessible/public anymore, such as `Locations`.

I used to rely on the `Locations` field to fetch the source code line number associated with a given instruction pointer. By using  `vm.Step()`, `vm.Position()` and the `Locations` of `vm.Program`, it allows to display debug/code to users when debugging their expression in [crowdsec](https://github.com/crowdsecurity/crowdsec).

With this field not being public anymore, how can I get the associated/relevant code for a given instruction pointer? 

Thanks!

**Comment by antonmedv** [comment]
Yes, I hid the location as assumed it was only used by errors. Let's revert it and make the location public. 

Note also what now `vm.Step()` will only work if expr built with `expr_debug` tag. See [debug.go](https://github.com/expr-lang/expr/blob/master/vm/debug.go). This change gives about 5% to 15% performance improvement. 

Also, for debugging expression by users it may be useful: https://expr-lang.org/expr-pro#explain

**Comment by buixor** [comment]
Hello,

>  Let's revert it and make the location public.

Thanks, I will be watching future releases. Do you want me to open an issue for traceability?

> Also, for debugging expression by users it may be useful: https://expr-lang.org/expr-pro#explain

Yes, I see that you have some cool features there. In [another discussion](https://github.com/expr-lang/expr/discussions/533), you mentioned having plans to make it available for FOSS. We're using expr in crowdsec that is MIT licensed, I'm interested to know more when you're ready!

**Reply by antonmedv** [reply, marked as answer]
> Do you want me to open an issue for traceability?

Yes, please.

**Reply by buixor** [reply]
https://github.com/expr-lang/expr/issues/546
